src/components/tl/mlx5/alltoall/alltoall.h

```diff
@@ -0,0 +1,195 @@
+   part of the algorithm. */
+typedef struct ucc_tl_mlx5_a2a_node {
+    int                  asr_rank;
+    ucc_sbgp_t *         sbgp;
+    void *               storage;
+    ucc_tl_mlx5_a2a_op_t ops[MAX_OUTSTANDING_OPS];
+    struct mlx5dv_mkey * team_recv_mkey;
+    void *               umr_entries_buf;
+    struct ibv_mr *      umr_entries_mr;
+} ucc_tl_mlx5_a2a_node_t;
+
+typedef struct a2a_net_ctrl {
+    struct {
+        void *   addr;
+        uint32_t rkey;
+    } atomic;
+    struct {
+        void *   addr;
+        uint32_t rkey;
+    } barrier;
+} a2a_net_ctrl_t;
+
+typedef struct ucc_tl_mlx5_a2a_net {
+    ucc_sbgp_t *      sbgp;
+    int               net_size;
+    int *             rank_map;
+    ucc_tl_mlx5_qp_t *rc_qps;
+    struct ibv_qp *   dct_qp;
+    struct ibv_srq *  srq;
+    uint32_t *        remote_dctns;
+    struct ibv_ah **  ahs;
+    struct ibv_cq *   cq;
+    struct ibv_cq *   umr_cq;
+    struct ibv_qp *   umr_qp;
+    struct ibv_mr *   ctrl_mr;
+    struct {
+#if ATOMIC_IN_MEMIC
+        struct ibv_dm *counters;
+#else
+        tl_mlx5_atomic_t *counters;
+#endif
+        struct ibv_mr *mr;
+    } atomic;
+    struct {
+        tl_mlx5_barrier_t *flags;
+        struct ibv_mr *    mr;
+    } barrier;
+    int *              blocks_sent; // net_size * MAX_OP_OUTSTANDING
+    a2a_net_ctrl_t *   remote_ctrl;
+    uint32_t *         rkeys;
+    ucc_tl_mlx5_dci_t *dcis;
+} ucc_tl_mlx5_a2a_net_t;
+
+typedef struct ucc_tl_mlx5_a2a_bcast_data {
+    int shmid;
+    int net_size;
+} ucc_tl_mlx5_a2a_bcast_data_t;
+
+enum
+{
+    TL_MLX5_A2A_STATE_SHMID,
+    TL_MLX5_A2A_STATE_EXCHANGE_PROGRESS,
+    TL_MLX5_A2A_STATE_EXCHANGE_DONE
+};
+
+typedef struct ucc_tl_mlx5_a2a {
+    struct ibv_pd *              pd;
+    struct ibv_context *         ctx;
+    int                          ib_port;
+    int                          state;
+    int                          transpose;
+    uint64_t                     max_msg_size;
+    ucc_rank_t                   team_size;
+    ucc_tl_mlx5_a2a_node_t       node;
+    ucc_tl_mlx5_a2a_net_t        net;
+    void *                       service_bcast_tmp_buf;
+    int                          sequence_number;
+    int                          op_busy[MAX_OUTSTANDING_OPS];
+    int                          cq_completions[MAX_OUTSTANDING_OPS];
+    int                          num_dci_qps;
+    uint8_t                      is_dc;
+    int                          previous_msg_size[MAX_OUTSTANDING_OPS];
+    void *                       previous_send_address[MAX_OUTSTANDING_OPS];
+    void *                       previous_recv_address[MAX_OUTSTANDING_OPS];
+    uint64_t                     dummy_atomic_buff;
+    int                          requested_block_size;
+    int                          max_num_of_columns;
+    struct ibv_mr *              scratch_bf_mr;
+    struct ibv_wc *              work_completion; // get rit
```

Comment:
remove comment?

Reply:
thanks for pointing out. I actually removed the field `work_completion` from the implementation